### PR TITLE
optimize readdir() open calls to be dealt with directly via 'fd'

### PR DESCRIPTION
**Pull request by harshavardhana**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
optimize readdir() open calls to be dealt with directly via 'fd'

## Motivation and Context
avoid poll.FD{} CPU issues

## How to test this PR?
Not easier

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Review by krisis — APPROVED** [review]

**Review by klauspost — COMMENTED** [review]
Neat!

`readDirWithOpts` should probably have the same change.

**Review by zveinn — APPROVED** [review]

**Comment by klauspost** [comment]
I am like 75% "let's add it" and "25% nah, too risky"...

File: cmd/os_unix.go [hunk 1]
```diff
@@ -39,6 +39,22 @@ func access(name string) error {
...
+// openFileWithFD return 'fd' based file descriptor
+func openFileWithFD(name string, flag int, perm os.FileMode) (fd int, err error) {
+	switch flag & writeMode {
+	case writeMode:
+		defer updateOSMetrics(osMetricOpenFileWFd, name)(err)
+	default:
+		defer updateOSMetrics(osMetricOpenFileRFd, name)(err)
+	}
+	var e error
+	fd, e = syscall.Open(name, flag, uint32(perm))
```
Comment (klauspost) [inline cmd/os_unix.go]:
```suggestion
	fd, e = syscall.Open(name, flag|syscall.O_CLOEXEC, uint32(perm))
```

**Review by klauspost — APPROVED** [review]
Maybe this.. Can't see it hurting us. But overall 👍🏼

**Pull request merged**